### ISSUE TITLE
Deadlock?

### ISSUE THREAD

**Issue by RouquinBlanc** [seanmonstar/spmc#5]
If I run the following loop, with 1 sender and 5 receiver, it happens that 1 or more threads never manage to finish.

```rust
extern crate spmc;

use std::thread;

fn main() {
    for l in 0..10 {
        println!("loop {}", l);

        let (tx, rx) = spmc::channel();

        let mut handles = Vec::new();
        for _ in 0..5 {
            let rx = rx.clone();
            handles.push(thread::spawn(move || {
                loop {
                    match rx.recv() {
                        Ok(_) => continue,
                        Err(_) => break,
                    }
                }
            }));
        }

        for i in 0..10 {
            tx.send(format!("Sending value {} {}", l, i)).unwrap();
        }
        drop(tx);

        for handle in handles {
            handle.join().unwrap();
        }
    }
}
```
The TX part is dropped, so all RX threads should finish quickly... There seems to be some kind of fundamental locking issue.

**Comment by seanmonstar** [comment]
You're right, the code was very naive. I've pushed to master a fix that properly handles the condvar notification, and also fixes #4's memory leak.

**Comment by RouquinBlanc** [comment]
No more leaking, no more blocking! thanks for the quick support!

**Comment by seanmonstar** [comment]
I've published 0.2.2 with the fixes.
